dvc/command/completion.py

```diff
@@ -0,0 +1,91 @@
+import argparse
+import logging
+import os
+
+import shtab
+
+from dvc.command import choices
+from dvc.command.base import CmdBaseNoRepo, append_doc_link
+
+logger = logging.getLogger(__name__)
+DEFAULT_FILENAME = {"bash": "dvc", "zsh": "_dvc"}
+CHOICE_FUNCTIONS = {
+    "bash": {"DVCFile": "_dvc_compgen_DVCFiles"},
+    "zsh": {"DVCFile": "_files -g '(*.dvc|Dvcfile)'"},
+}
+PREAMBLE = {
+    "bash": """
+# $1=COMP_WORDS[1]
+_dvc_compgen_DVCFiles() {
+  compgen -f -X '!*?.dvc' -- $1  # DVC-files
+  compgen -d -S '/' -- $1  # recurse into subdirs
+}
+""",
+    "zsh": "",
+}
+
+
+class CmdCompletion(CmdBaseNoRepo):
+    def run(self):
+        from dvc.cli import get_main_parser
+
+        parser = get_main_parser()
+        shell = self.args.shell
+        script = shtab.complete(
+            parser,
+            shell=shell,
+            preamble=PREAMBLE[shell],
+            choice_functions=CHOICE_FUNCTIONS[shell],
+        )
+
+        if self.args.dir == "-":
+            logger.debug("Writing tab completion to stdout")
+            print(script)
+        else:
+            fname = os.path.join(
+                self.args.dir, self.args.file or DEFAULT_FILENAME[shell],
+            )
+            logger.info(f"Writing tab completion to {fname}")
+            with open(fname, "w") as fd:
+                print(script, file=fd)
+        return 0
+
+
+def add_parser(subparsers, parent_parser):
+    COMPLETION_HELP = "Install shell tab completion."
+    COMPLETION_DESCRIPTION = (
+        "Automatically generates shell tab completion scripts."
+    )
+    completion_parser = subparsers.add_parser(
+        "completion",
+        parents=[parent_parser],
+        description=append_doc_link(COMPLETION_DESCRIPTION, "completion"),
+        help=COMPLETION_HELP,
+        formatter_class=argparse.RawDescriptionHelpFormatter,
+    )
+    completion_parser.add_argument(
+        "-s",
+        "--shell",
+        help="Shell syntax for completions.",
+        default="bash",
+        choices=["bash", "zsh"],
+    )
+    completion_parser.add_argument(
+        "-f",
+        "--file",
+        help=(
+            "File name for output. Defaults depending on --shell:"
+            f" {DEFAULT_FILENAME}."
+        ),
+    )
+    completion_parser.add_argument(
+        "dir",
+        help=(
+            "Output directory for completion script."
+            " Defaults to '-' for stdout (ignoring --file)."
+        ),
+        default="-",
+        nargs="?",
+        choices=choices.Required.DIR,
```

Comment:
Doesn't look like it is required :) Btw, can't we distinguish required/optional from nargs?

Reply:
it is required. A bit of cleverness to get the desired behaviour.

- `nargs=?` allows no user-specified arg, resulting in default=-, ergo there is still a filled in. Forcefully unsetting (e.g. user providing empty string) will result in an error because this arg is very much required
- `choices.(Required|Optional)` is merely an extra error check and does correctly describe the fact that this is a required arg

We could use `choices.Optional` to essentially avoid `shtab` asserting non-null values, but this way seems nicer.

Reply:
I'm not sure, but maybe removing nargs would pretty much result in the same desired behaviour without causing this confusion?

Reply:
no - removing is the same as saying it's required to be explicitly specified. I'd prefer `nargs='?', default='-'` as the exact user-intuitive behaviour needed.